### ISSUE TITLE
Why does enabling gradient accumulation cause GRPO performance to drop？

### ISSUE THREAD

**Issue by Alex-Songs** [huggingface/trl#3021]
Why does enabling gradient accumulation cause GRPO performance to drop, and why does performance decrease as accumulation increases?

**Comment by qgallouedec** [comment]
Can you share some code and results?

**Comment by JosephChenHub** [comment]
I noticed that the updated version is  ```loss = ((per_token_loss * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)).mean()```

maybe it has been resolved.

**Comment by loxs123** [comment]
> I noticed that the updated version is  ```loss = ((per_token_loss * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)).mean()```
> 
> maybe it has been resolved. 

Have you tried the modified effect?

**Comment by loxs123** [comment]
> I noticed that the updated version is `loss = ((per_token_loss * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)).mean()`
> 
> maybe it has been resolved.
```python
if self.loss_type == "grpo":
    loss = ((per_token_loss * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)).mean()
elif self.loss_type == "bnpo":
    loss = (per_token_loss * completion_mask).sum() / completion_mask.sum().clamp(min=1.0)
elif self.loss_type == "dr_grpo":
    loss = (per_token_loss * completion_mask).sum() / (per_token_loss.size(0) * self.max_completion_length)
else:
    raise ValueError(f"Unknown loss type: {self.loss_type}")
```

The complete code is shown above, so the issue you just mentioned likely doesn’t exist for `grpo_loss`/`dr_grpo`. However, the `bnpo` (and probably `dapo` as well) loss still seems to have the issue. Additionally, `grpo_loss` and `bnpo_loss` are inherently two different loss calculation methods. I believe that `loss = ((per_token_loss * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)).mean()` is not a correction to the original loss, but rather follows the original GRPO algorithm.

**Comment by albertvillanova** [comment]
Hey @Alex-Songs, sorry for the very slow reply — we're catching up on the backlog. 

What you (and @JosephChenHub / @loxs123 in the thread) observed is the classic loss-normalization-under-gradient-accumulation effect: when the loss is normalized by the token count of each *local* accumulation micro-batch, the micro-batches get mis-weighted once their gradients are averaged, and the mismatch grows with more accumulation steps — exactly why it worsened as you raised accumulation. 

On current main the default `loss_type` is `dapo`, which normalizes by the active-token count of the *global accumulated batch* (across all micro-batches and processes), so the loss is invariant to the number of gradient-accumulation steps — see https://github.com/huggingface/trl/blob/7b4c3b3b208c8611ff23818e8483cd2a6eaecb35/trl/trainer/grpo_trainer.py#L2728-L2730 and https://github.com/huggingface/trl/blob/7b4c3b3b208c8611ff23818e8483cd2a6eaecb35/trl/trainer/grpo_trainer.py#L1878 (`grpo` and `dr_grpo` are GA-robust too). 

Closing as fixed — please re-open with your config and `trl env` if you still see drift with the default config on a recent version. One caveat: `loss_type="bnpo"` normalizes per local batch by design and isn't accumulation-invariant, so prefer the default `dapo`.
